lib/auth/grpcserver.go

```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
 	return nil
 }
 
-func validateUserSingleUseCertRequest(req *proto.UserCertsRequest, clock clockwork.Clock) error {
+var errUserSingleUseUserCertNotNeeded = errors.New("single-use user certificate not needed to access this resource")
+
+// validateUserSingleUseCertRequest validates the request for a single-use user
+// cert. Returns nil if validation succeeds and MFA check is required, or
+// errUserSingleUseUserCertNotNeeded if validation succeeds and MFA
+// authentication is not required for this session. Any other error means
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
+		return trace.BadParameter("cannot request a single-use certificate for a different user")
+	}
+
+	var noMFAAccessErr error
 	switch req.Usage {
 	case proto.UserCertsRequest_SSH:
 		if req.NodeName == "" {
 			return trace.BadParameter("missing NodeName field in a ssh-only UserCertsRequest")
 		}
+		// Find the target node and check whether MFA is required.
+		nodes, err := actx.authServer.GetNodes(defaults.Namespace, services.SkipValidation())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		var matches []types.Server
+		for _, n := range nodes {
+			// Get the server address without port number.
+			addr, _, err := net.SplitHostPort(n.GetAddr())
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				matches = append(matches, n)
+			}
+		}
+		if len(matches) == 0 {
+			return trace.NotFound("target resource not found")
```

Comment:
Is the resource type/name not revealed in these messages on purpose?

Reply:
I was just being lazy, updated